### DISCUSSION TITLE
Modernize FreshRSS proposal

### DISCUSSION THREAD

**Discussion by yguedidi** [FreshRSS/FreshRSS#4322] [General]
Hello!

I'd like to work on an open source project. And I like refactoring and modernizing code from legacy.
FreshRSS is a good project, but from my point of view its codebase looks legacy.
So I'd like to propose my help to modernize it.

Here main improvement ideas:

- follow [PHP The Right way](https://phptherightway.com/) principles (Composer, ...)
- follow [SOLID](https://dev.to/evrtrabajo/solid-in-php-d8e) principles (no side effect code, close API, ...)
- follow [Clean Code](https://github.com/jupeter/clean-code-php) principles (more readable code, ...)
- follow [Twelve-Factor App](https://12factor.net/) principles (environment variables, ...)
- use latest PHP (8.0 or even 8.1)
- use trusted vendors: Doctrine for DB, Monolog for logs, a templating language (Twig), certainly others
- more tests
- use modern front stack (SCSS, Webpack, ...)

I know it's a huge list!
If it's something you can envision for FreshRSS future, I'd more more than happy to help going in that direction!
Maybe on a new `next` branch or using the existing `edge` one.
This is the best option for FreshRSS evolution IMO.

If your not aligned with this proposal, I will respect that.
Will you accept then that I fork FreshRSS and I rebrand it into something else, so that I can play with refactoring and modernization?
Of course the new project based on current Fresh RSS will be under the same AGPL licence.

Thank you

**Comment by Frenzie** [comment]
> I know it's a huge list!

The first four items seem to be descriptive. Could you please be more concrete?

As an aside, note that this is what an actual argument for Composer looks like: "Doctrine for DB, Monolog for logs, a templating language (Twig), certainly others." Paraphrased, I want to add a lot more libraries and Composer would facilitate that. That's quite different than saying you want to add Composer pretty much completely at random.

> use latest PHP (8.0 or even 8.1)

See https://github.com/FreshRSS/FreshRSS/discussions/3321

> use trusted vendors: Doctrine for DB, Monolog for logs, a templating language (Twig), certainly others

I do like Django syntax.

> more tests

Definitely.

**Comment by Alkarex** [comment]
Hello,
You are welcome to experiment and report your progress, but with so many changes, it is probably just as good to restart from scratch.
On my side, I would rather like to spend my efforts in the near future on adding some of the features I am interested in, and I do not see the current code base as something that slows me down.
We do have some light code improvement tasks open, such as https://github.com/FreshRSS/FreshRSS/issues/4112

**Reply by Alkarex** [reply]
However, it would be possible to carve a part that you could focus on, and which could be incorporated. This could for instance be a JavaScript-based UI (talking to more or less our existing back-end) as an alternative to our current mostly-server-side UI (I would personally welcome such an effort), or another server-side feature in form of library.

**Reply by Alkarex** [reply]
@yguedidi If you could be interested in experimenting with a new alternative front-end, that would be excellent, and I could assist. It is something I had on my own todo for the future.

**Reply by yguedidi** [reply]
@Alkarex what your plans for front end? just curious, not sure I'll take it, I'm more PHP :sweat_smile: 
but could be a could challenge!

**Comment by yguedidi** [comment]
@Art4 what do you think about this?
